beacon-chain/forkchoice/doubly-linked-tree/forkchoice.go

```diff
@@ -714,3 +714,114 @@ func (f *ForkChoice) ParentRoot(root [32]byte) ([32]byte, error) {
 	}
 	return n.parent.root, nil
 }
+
+// OnPayloadAttestationMessage processes a new aggregated payload attestation message.
+func (s *Store) OnPayloadAttestationMessage(
+	ctx context.Context,
+	payloadAttestation *ethpb.PayloadAttestation,
+	isFromBlock bool) error {
+	data := payloadAttestation.Data
+	blockRoot := bytesutil.ToBytes32(data.BeaconBlockRoot)
+
+	// Check if the block exists in the store
+	node, ok := s.nodeByRoot[blockRoot]
+	if !ok {
+		return errors.New("beacon block root unknown")
+	}
+
+	// Only update payload boosts with attestations from a block if the block is for the current slot and it's early
+	if isFromBlock {
+		currentSlot := slots.CurrentSlot(s.genesisTime)
+		if data.Slot+1 != currentSlot {
+			return nil
+		}
+		timeIntoSlot := (uint64(time.Now().Unix()) - s.genesisTime) % params.BeaconConfig().SecondsPerSlot
+		if timeIntoSlot >= params.BeaconConfig().SecondsPerSlot/params.BeaconConfig().IntervalsPerSlot {
+			return nil
+		}
+	}
+
+	// Update the ptc vote for the block
+	ptcVote := s.ptcVote[blockRoot]
+	if ptcVote == nil {
+		ptcVote = make([]primitives.PTCStatus, fieldparams.PTCSize)
+	}
+
+	for i, vote := range payloadAttestation.AggregationBits {
+		if vote == byte(0) {
+			ptcVote[i] = data.PayloadStatus
+		}
+	}
+	s.ptcVote[blockRoot] = ptcVote
+
+	// Update the payload boosts if threshold has been achieved
+	presentCount := 0
+	withheldCount := 0
+	for _, vote := range ptcVote {
+		if vote == primitives.PAYLOAD_PRESENT {
+			presentCount++
+		} else if vote == primitives.PAYLOAD_WITHHELD {
+			withheldCount++
+		}
+	}
+
+	if presentCount > int(params.BeaconConfig().PayloadTimelyThreshold) {
+		s.payloadRevealBoostRoot = blockRoot
+	}
+	if withheldCount > int(params.BeaconConfig().PayloadTimelyThreshold) {
+		s.payloadWithholdBoostRoot = node.parent.root
+		s.payloadWithholdBoostFull = s.isParentNodeFull(node)
+	}
+
+	// Update the PTC vote for the block
+	s.updatePTCVote(blockRoot, payloadAttestation)
+
+	// Update payload boosts if necessary
+	s.updatePayloadBoosts(blockRoot, node)
+
+	return nil
+}
+
+func (s *Store) updatePTCVote(blockRoot [32]byte, payloadAttestation *ethpb.PayloadAttestation) {
```

Comment:
Why is the logic of this function being repeated in the main handler?

Reply:
forgot to remove code, thanks!